File: src/main/java/com/refinedmods/refinedstorage/apiimpl/util/Comparer.java
package com.refinedmods.refinedstorage.apiimpl.util;

import com.refinedmods.refinedstorage.api.util.IComparer;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.fluids.FluidStack;

import javax.annotation.Nonnull;

public class Comparer implements IComparer {
    @Override
    public boolean isEqual(@Nonnull ItemStack left, @Nonnull ItemStack right, int flags) {
        if (left.isEmpty() && right.isEmpty()) {
            return true;
        }

        if (!ItemStack.isSame(left, right)) {
            return false;
        }

        if ((flags & COMPARE_NBT) == COMPARE_NBT && !ItemStack.tagMatches(left, right)) {
            return false;
        }

        if ((flags & COMPARE_QUANTITY) == COMPARE_QUANTITY && left.getCount() != right.getCount()) {
            return false;
        }

        return true;
    }

    @Override
    public boolean isEqual(@Nonnull FluidStack left, @Nonnull FluidStack right, int flags) {
        if (left.isEmpty() && right.isEmpty()) {
            return true;
        }

        if (left.getFluid() != right.getFluid()) {
            return false;
        }

        if ((flags & COMPARE_NBT) == COMPARE_NBT && !FluidStack.areFluidStackTagsEqual(left, right)) {
            return false;
        }

        if ((flags & COMPARE_QUANTITY) == COMPARE_QUANTITY && left.getAmount() != right.getAmount()) {
            return false;
        }

        return true;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/util/Comparer.kt
package com.refinedmods.refinedstorage.apiimpl.util

import com.refinedmods.refinedstorage.api.util.IComparer
import net.minecraft.item.ItemStack
import reborncore.common.fluid.container.FluidInstance


class Comparer : IComparer {
    override fun isEqual(left: ItemStack, right: ItemStack, flags: Int): Boolean {
        if (left.isEmpty && right.isEmpty) {
            return true
        }
        if (!ItemStack.areItemsEqual(left, right)) {
            return false
        }
        if (flags and IComparer.COMPARE_NBT == IComparer.COMPARE_NBT) {
            if (!ItemStack.areTagsEqual(left, right)) {
                return false
            }
        }
        if (flags and IComparer.COMPARE_QUANTITY == IComparer.COMPARE_QUANTITY) {
            if (left.count != right.count) {
                return false
            }
        }
        return true
    }

    override fun isEqual(left: FluidInstance, right: FluidInstance, flags: Int): Boolean {
        if (left.isEmpty() && right.isEmpty()) {
            return true
        }
        if (left.getFluid() !== right.getFluid()) {
            return false
        }
        if (flags and IComparer.COMPARE_NBT == IComparer.COMPARE_NBT) {
            if (left.tag != right.tag) {
                return false
            }
        }
        if (flags and IComparer.COMPARE_QUANTITY == IComparer.COMPARE_QUANTITY) {
            if (left.getAmount() !== right.getAmount()) {
                return false
            }
        }
        return true
    }
}